adr/001-component-library.md

```diff
@@ -0,0 +1,65 @@
+# Introduce component library
+
+- Status: Accepted
+- Deciders: Team
+- Date: 17.10.2024
+
+## Result
+
+A1: Component library is introduced, firstly as just a folder in our current setup, adding yarn workspaces or similar requires more research and testing
+
+## Problem context
+
+Today our UI components are tightly coupled to the app in which they are rendered.
+
+This leads to several issues:
+
+- Makes it hard to do component testing outside a fully rendered app.
+  - See tests in `test/e2e/integration` for examples of how we have most of our tests in integration tests because components need to be running inside an app to work.
+  - See `src/test/renderWithProviders.tsx` for examples of how much scaffolding we need to do to run component tests currently. This also leads to very slow component tests.
+- Makes refactoring the app framework a lot harder.
+  - If we search for `useNodeItem` in src/layout we get over 100 hits. If we make changes or remove hooks like this, every single UI component must be updated.
+- Leads to unclear interfaces between UI components and the app framework.
+  - See: `src/layout/Button/ButtonComponent.tsx`. This component references `node.parent` which confuses the role and responsibility of the button component.
+- Makes developing UI components complex without deep understanding of the application.
+- Enables sharing of pure components (docs, Studio, storybook)
+
+## Decision drivers
+
+A list of decision drivers. These are points which can differ in importance. If a point is "nice to have" rather than
+"need to have", then prefix the description.
+
+- B1: UI components should only receive data to display, and notify the app when data is changed.
+- B2: UI components should live in a separate folder from the app itself, and have no dependencies to the app.
```

Comment:
This is like saying its confusing to use designsystemet as a view library.

The benefits of separating view components from the app logic includes but is not limited to:

**1. Code Reusability:** Pure view components are reusable across different parts of the app, as they are only concerned with how the UI looks, not how it behaves. This makes it easy to use the same component in multiple places without duplicating logic.

This is a more important point than reusing components with Studio or Storybook, but it also gives us that possibility.

**2. Simplified Testing:** With view components separated, you can test them independently without worrying about application logic. This isolation makes unit tests more focused and reliable, as you only need to check if a component renders correctly for given inputs (props) and not the entire app flow.

This is a huge issue in our app currently. This will ensure we can both move more testing to components which will be faster now and scale better in the future.

**3. Easier Maintenance:** Separation makes it easier to locate and fix bugs since you know exactly where the UI ends and the logic begins. This modularity also makes refactoring or redesigning the UI simpler, as you’re only changing visual components without touching business logic.

As we are reusing things like lists, tables, inputs etc, we only update in one spot as well.

**4. Improved Readability:** Clean separation between views and logic helps other developers (or your future self) understand the structure of the app more quickly. 

They can identify what each part does at a glance, as there is a clear boundary between the UI layer and the functionality layer.

**5. Scalability:** As applications grow, so does their complexity. A separation between UI and logic allows for better scalability by organizing code into distinct concerns. This allows teams to work on specific parts of the application more independently, making collaboration smoother.

In the future it could make sense to have one team focused on input components, and another focused on the app logic.

**Easier State Management:** When the view components are separate, managing state becomes clearer. Logic components handle state, business rules, and data fetching, while view components remain “dumb” and only render the provided data. This results in a more predictable data flow.